common/ensemble_building/ensemble_selection.py

```diff
@@ -0,0 +1,210 @@
+    ) -> None:
+        self.ensemble_size = ensemble_size
+        self.loss_fn: Optional[Callable[..., float]] = loss_fn
+        self.loss_fn_args = loss_fn_args
+        self.random_state = random_state
+
+    def __getstate__(self) -> Dict[str, Any]:
+        # Cannot serialize a metric if
+        # it is user defined.
+        # That is, if doing pickle dump
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
+        self.loss_fn = None
+        self.loss_fn_args = {}
+        return self.__dict__
+
+    def fit(
+        self,
+        predictions: List[np.ndarray],
+        labels: np.ndarray,
+        identifiers: List[Tuple[int, int, float]],
+    ) -> AbstractEnsemble:
+        self.ensemble_size = int(self.ensemble_size)
+        if self.ensemble_size < 1:
+            raise ValueError("Ensemble size cannot be less than one!")
+
+        self._fit(predictions, labels)
+        self._calculate_weights()
+        self.identifiers_ = identifiers
+        return self
+
+    def _fit(
+        self,
+        predictions: List[np.ndarray],
+        labels: np.ndarray,
+    ) -> AbstractEnsemble:
+        """Fast version of Rich Caruana's ensemble selection method."""
+
+        if self.loss_fn is None:
+            raise ValueError(
+                "A function with signature Callable[[labels, predictions, ...], float]"
+                "is needed to perform ensemble selection."
+            )
+
+        self.num_input_models_ = len(predictions)
+
+        ensemble: List[np.ndarray] = []
+        trajectory: List[float] = []
+        order: List[int] = []
+
+        ensemble_size = self.ensemble_size
+
+        weighted_ensemble_prediction = np.zeros(
+            predictions[0].shape,
+            dtype=np.float64,
+        )
+        fant_ensemble_prediction = np.zeros(
+            weighted_ensemble_prediction.shape,
+            dtype=np.float64,
+        )
+        for _ in range(ensemble_size):
+            losses = np.zeros(
+                (len(predictions)),
+                dtype=np.float64,
+            )
+            s = len(ensemble)
+            if s > 0:
+                np.add(
+                    weighted_ensemble_prediction,
+                    ensemble[-1],
+                    out=weighted_ensemble_prediction,
+                )
+
+            # Memory-efficient averaging!
+            for j, pred in enumerate(predictions):
+                # fant_ensemble_prediction is the prediction of the current ensemble
+                # and should be ([predictions[selected_prev_iterations] + predictions[j])/(s+1)
+                # We overwrite the contents of fant_ensemble_prediction
+                # directly with weighted_ensemble_prediction + new_prediction and then scale for avg
+                np.add(weighted_ensemble_prediction, pred, out=fant_ensemble_prediction)
```

Comment:
Only this part (line 99) is memory-efficient and it is worth keeping the current form.